kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceRateLimiter.java

```diff
@@ -98,29 +105,48 @@ private boolean isOverGracePeriod(Long nowMs) {
     return false;
   }
 
-  private int addAndGetRate(long nowMs) {
-    rateCounter.add(nowMs);
+  private void addToRateLimiter(TimeAndSize thisMessage) {
+    rateCounter.add(thisMessage);
     rateCounterSize.incrementAndGet();
+    byteCounterSize.addAndGet(thisMessage.size);
 
     synchronized (rateCounter) {
-      if (rateCounter.peekLast() < nowMs - ONE_SECOND_MS) {
+      if ((rateCounter.peekLast().time < thisMessage.time - ONE_SECOND_MS)) {
         rateCounter.clear();
         rateCounterSize.set(0);
       } else {
-        while (rateCounter.peek() < nowMs - ONE_SECOND_MS) {
-          rateCounter.poll();
+        while (rateCounter.peek().time < thisMessage.time - ONE_SECOND_MS) {
+          TimeAndSize messageToRemove = rateCounter.poll();
+          byteCounterSize.addAndGet(-messageToRemove.size);
           rateCounterSize.decrementAndGet();
         }
       }
     }
-    return rateCounterSize.get();
+    return;
   }
 
-  private Optional<Duration> getWaitFor(int currentRate) {
-    if (currentRate <= maxRequestsPerSecond) {
+  private Optional<Duration> getWaitFor(int currentCountRate, long currentByteRate) {
+    if (currentCountRate <= maxRequestsPerSecond && currentByteRate <= maxBytesPerSecond) {
       return Optional.empty();
     }
-    double waitForMs = ((double) currentRate / (double) maxRequestsPerSecond - 1) * 1000;
+
+    double waitForMs;
+    if (currentCountRate > maxRequestsPerSecond) {
+      waitForMs = ((double) currentCountRate / (double) maxRequestsPerSecond - 1) * 1000;
+    } else {
+      waitForMs = ((double) currentByteRate / (double) maxBytesPerSecond - 1) * 1000;
+    }
+
     return Optional.of((Duration.ofMillis((long) waitForMs)));
   }
+
+  private static final class TimeAndSize {
+    private long time;
+    private long size;
```

Comment:
nit: It looks like these should have been `final`. Them being effectively `final` should reap the same benefits with contemporary JVMs though.